storage/paths/db_index.go

```diff
@@ -0,0 +1,891 @@
+		for rft := range read.AllSet() {
+			if lockMap[storiface.SectorFileType(rft)].writeTs.Valid &&
+				lockMap[storiface.SectorFileType(rft)].writeTs.Time.After(time.Now().Add(LockTimeOut)) {
+				return false, xerrors.Errorf("Cannot acquire read lock for sector %v filetype %d already locked for writing", sector, rft)
+			}
+		}
+
+		// Acquire write locks
+		_, err = dbi.harmonyDB.Exec(ctx,
+			`UPDATE sectorlocation
+							 	SET write_ts = NOW()
+							 	WHERE miner_id=$1
+							 		AND sector_num=$2
+							 		AND sector_filetype = ANY($3)`,
+			sector.Miner, sector.Number, write.AllSet())
+		if err != nil {
+			return false, xerrors.Errorf("Acquiring write locks for sector %v fails with err: %v", sector, err)
+		}
+
+		// Acquire read locks
+		_, err = dbi.harmonyDB.Exec(ctx,
+			`UPDATE sectorlocation
+							 	SET read_ts = NOW(), read_refs = read_refs + 1
+							 	WHERE miner_id=$1
+							 		AND sector_num=$2
+							 		AND sector_filetype = ANY($3)`,
+			sector.Miner, sector.Number, read.AllSet())
+		if err != nil {
+			return false, xerrors.Errorf("Acquiring read locks for sector %v fails with err: %v", sector, err)
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (dbi *DBIndex) StorageLock(ctx context.Context, sector abi.SectorID, read storiface.SectorFileType, write storiface.SectorFileType) error {
+
+	_, err := dbi.lock(ctx, sector, read, write)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (dbi *DBIndex) StorageTryLock(ctx context.Context, sector abi.SectorID, read storiface.SectorFileType, write storiface.SectorFileType) (bool, error) {
+	return dbi.lock(ctx, sector, read, write)
+}
+
+func (dbi *DBIndex) StorageGetLocks(ctx context.Context) (storiface.SectorLocks, error) {
+
+	var rows []struct {
+		MinerId        uint64
+		SectorNum      uint64
+		SectorFileType int
+		ReadTs         sql.NullTime
+		ReadRefs       int
+		WriteTs        sql.NullTime
+	}
+
+	err := dbi.harmonyDB.Select(ctx, &rows,
+		"SELECT miner_id, sector_num, sector_filetype, read_ts, read_refs, write_ts FROM sectorlocation")
+	if err != nil {
+		return storiface.SectorLocks{}, err
+	}
+
+	type locks struct {
+		sectorFileType storiface.SectorFileType
+		readRefs       uint
+		writeLk        bool
+	}
+	sectorLocks := make(map[abi.SectorID]locks)
+	for _, row := range rows {
+		sector := abi.SectorID{
+			Miner:  abi.ActorID(row.MinerId),
+			Number: abi.SectorNumber(row.SectorNum),
+		}
+		sectorLocks[sector] = locks{
+			sectorFileType: storiface.SectorFileType(row.SectorFileType),
+			readRefs:       uint(row.ReadRefs),
+			writeLk:        row.WriteTs.Valid,
+		}
+	}
+
+	var result storiface.SectorLocks
+	for sector, locks := range sectorLocks {
+		var lock storiface.SectorLock
+		lock.Sector = sector
+		lock.Read[locks.sectorFileType] = locks.readRefs
+		if locks.writeLk {
+			lock.Write[locks.sectorFileType] = 1
+		} else {
+			lock.Write[locks.sectorFileType] = 0
+		}
```

Comment:
* We don't assign to the result
* Should probably filter out timed-out locks here

Reply:
done